kong/plugins/oauth2/access.lua

```diff
@@ -149,12 +149,20 @@ local function authorize(conf)
       if not allowed_redirect_uris then
         response_params = {[ERROR] = "invalid_client", error_description = "Invalid client authentication" }
       else
-        redirect_uri = parameters[REDIRECT_URI] and parameters[REDIRECT_URI] or allowed_redirect_uris[1]
-
-        if not utils.table_contains(allowed_redirect_uris, redirect_uri) then
-          response_params = {[ERROR] = "invalid_request", error_description = "Invalid " .. REDIRECT_URI .. " that does not match with any redirect_uri created with the application" }
-          -- redirect_uri used in this case is the first one registered with the application
-          redirect_uri = allowed_redirect_uris[1]
+        -- Since default redirect URIs are now patterns, it's not safe to use them directly before the
+        -- comparison below  as before. Instead, if one is not supplied, just use the one registered with
+        -- the application. In either case, we should try to remove the start/end of string characters.
+        -- This is pretty hacky, perhaps a "default redirect URI" option should be provided instead.
+        redirect_uri = parameters[REDIRECT_URI]
+
+        if redirect_uri then
+          if not utils.table_contains_regex(allowed_redirect_uris, redirect_uri) then
+            response_params = {[ERROR] = "invalid_request", error_description = "Invalid " .. REDIRECT_URI .. " that does not match with any redirect_uri created with the application" }
+            -- redirect_uri used in this case is the first one registered with the application
+            redirect_uri = string.gsub(allowed_redirect_uris[1],"[%^%$]","")
```

Comment:
You're right if a user has indicated that the first URI be a more complex regex that this would fail. I'm not sure of a better way to do this than to simply not handle things this way but I was hoping for a more in depth discussion of how to handle this and the comment about the "hackiness" of this behavior.

In our current OAuth AS, we have a "Client Home URL" option that's used as the default redirect URI when none is supplied by the client. If we do this here, we'd also need to support that in the migration, which isn't a big deal.

I'm trying to avoid overcomplicating this too much if possible although I understand that the behavior as implemented is unexpected and confusing.